api/src/Rbac/Access/Gate.php

```diff
@@ -0,0 +1,237 @@
+            if (! empty($this->defaultRoles) && $this->checkAccessRecursive($itemName, $arguments)) {
+                return true;
+            }
+
+            $user = $this->resolveUser();
+        } catch (UserNotFoundException $e) {
+            return false;
+        }
+
+        $assignments = $this->getStorage()->getAssignments($user->getAuthIdentifier());
+
+        return $this->checkAccessRecursive($itemName, $arguments, $assignments);
+    }
+
+    /**
+     * Get Gate service for particular user.
+     *
+     * @param Authenticatable $user
+     * @return static
+     */
+    public function forUser(Authenticatable $user)
+    {
+        return new static($this->getStorage(), function () use ($user) {return $user;}, $this->defaultRoles);
+    }
+
+    /**
+     * Check permissions and roles recursively.
+     *
+     * @param $itemName
+     * @param $params
+     * @param array $assignments
+     * @param array $itemStack
+     * @return bool
+     */
+    protected function checkAccessRecursive($itemName, $params, $assignments = [], $itemStack = [])
+    {
+        if (! ($item = $this->getItem($itemName))) {
+            return false;
+        }
+
+        $itemStack[$item->name] = $item;
+
+        if (isset($assignments[$itemName]) || in_array($itemName, $this->defaultRoles)) {
+            foreach ($itemStack as $itemName => $item) {
+                if (! $this->executeRule($item, $params)) {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        foreach ($this->getStorage()->getParentNames($itemName) as $parent) {
+            if ($this->checkAccessRecursive($parent, $params, $assignments, $itemStack)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * @param $itemName
+     * @return Item|false
+     */
+    protected function getItem($itemName)
+    {
+        if (! isset(static::$itemCache[$itemName])) {
+            static::$itemCache[$itemName] = $this->getStorage()->getItem($itemName);
+        }
+
+        return static::$itemCache[$itemName];
+    }
+
+    /**
+     * @param Item $item
+     * @param $params
+     * @return bool
+     */
+    protected function executeRule(Item $item, $params)
+    {
+        if ($item->getRuleName() !== null) {
+            $rule = $this->getRule($item->getRuleName());
+            if (! $rule->execute($this->userResolver, $params)) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    /**
+     * @param $ruleName
+     * @return Rule
+     */
+    protected function getRule($ruleName)
+    {
+        if (! isset(static::$ruleCache[$ruleName])) {
+            static::$ruleCache[$ruleName] = new $ruleName;
+        }
+
+        return static::$ruleCache[$ruleName];
+    }
+
+    /**
+     * @return StorageInterface
+     */
+    public function getStorage()
+    {
+        return $this->storage;
```

Comment:
storage is a facade now which includes itemsStorage and assignmentStrogae either can be file or database based